[Russell Saerang] iP

RussellDash332 opened the pull request:
## [Stashy](https://www.youtube.com/watch?v=dQw4w9WgXcQ)
> Cannot come up with a name - Russell, 2022

Stash it, get it done!
- easy to use
- easy to not use
- very ~~1337~~ COOL message, _trust me!_

Use this application by setting it up in **under one minute** (~~unless bad WiFi~~)
1. go to [this page](https://github.com/RussellDash332/ip/releases/tag/A-Jar)
2. read the instructions there
3. have fun! 🥳 

Current features include:
- [X] Support _**basic**_ operations such as add, delete, find
- [x] @RussellDash332's wishlist to support multiple tasks in one go
- [ ] Literally anything else

The application mainly uses Java, so here's what a Java code looks like.
```java
public class Main { // wow syntax highlighting!
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
Maybe a little bit of shell script?
```sh
echo $PATH
```

Commit:
Add Gradle support

RussellDash332 committed:
[Level-1] Greet, Echo, Exit

RussellDash332 committed:
[Level-2] Add, List

RussellDash332 committed:
[Level-3] Mark as Done

RussellDash332 committed:
Add Deadlines class

RussellDash332 committed:
Add Events class

RussellDash332 committed:
Add ToDos class

RussellDash332 committed:
[Level-4] ToDo, Event, Deadline

RussellDash332 committed:
Fix Checkstyle issues

RussellDash332 committed:
Add indentation and horizontal line

RussellDash332 committed:
Modify gitignore

RussellDash332 committed:
Modify runtest scripts and their permission

RussellDash332 committed:
[A-TextUiTesting] Automated Text UI Testing

RussellDash332 committed:
Add DukeException class

RussellDash332 committed:
Incorporate DukeException on main code

RussellDash332 committed:
Modify UI testing

RussellDash332 committed:
Modify new expected UI text

RussellDash332 committed:
Modify code abstraction to be more concise

RussellDash332 committed:
Add Javadocs

RussellDash332 committed:
[Level-5] Handle Errors

RussellDash332 committed:
Modify UI test input and expected result

RussellDash332 committed:
[Level-6] Delete

RussellDash332 committed:
Modify name and message

RussellDash332 committed:
Handle more input scenarios

RussellDash332 committed:
Modify test case for UI testing

RussellDash332 committed:
Add Javadoc comments

RussellDash332 committed:
[Level-7] Save

RussellDash332 committed:
Add save functionality

RussellDash332 committed:
[Level-8] Dates and Times

RussellDash332 committed:
Merge branch 'branch-Level-7'
Create pull request from branch-Level-7 to master

RussellDash332 committed:
Merge branch 'branch-Level-8'
Create pull request from branch-Level-8 to master

RussellDash332 committed:
Rename Duke to Stashy

RussellDash332 committed:
Modify gitignore

RussellDash332 committed:
Add Storage

RussellDash332 committed:
Rename ToDo, Event, and Deadline

RussellDash332 committed:
Add Javadoc comment

RussellDash332 committed:
Fix mismatching type

RussellDash332 committed:
Add UI

RussellDash332 committed:
Make Task class abstract

RussellDash332 committed:
Move custom error message to UI

RussellDash332 committed:
Move TaskReader fully to Storage

RussellDash332 committed:
Modify Deadline and Event datetime format

RussellDash332 committed:
Modify gitignore

RussellDash332 committed:
Modify text UI test

RussellDash332 committed:
Add Javadoc comment and fix datetime format

RussellDash332 committed:
Change method signatures

RussellDash332 committed:
Allow overloading for Task class

RussellDash332 committed:
[A-MoreOOP] Use More OOP

RussellDash332 committed:
Move files to new directories

RussellDash332 committed:
Rename java directory to stashy

RussellDash332 committed:
[A-Packages] Organize into Packages

RussellDash332 committed:
Merge branch 'add-gradle-support' of github.com:nus-cs2103-AY2223S1/ip

RussellDash332 committed:
[A-Gradle] Use Gradle

RussellDash332 committed:
Add JUnit methods

RussellDash332 committed:
Modify build.gradle

RussellDash332 committed:
Move stashy to java/stashy

RussellDash332 committed:
[A-JUnit] Add JUnit Tests

RussellDash332 committed:
[A-Jar] Create a JAR File

RussellDash332 committed:
Add JavaDoc comments to Command classes

RussellDash332 committed:
Add JavaDoc comments to Task and exception classes

RussellDash332 committed:
Add JavaDoc comments for unit test classes

RussellDash332 committed:
[A-JavaDoc] JavaDoc

RussellDash332 committed:
Merge branch 'branch-A-JavaDoc'

RussellDash332 committed:
[A-CodingStandard] Follow the Coding Standard

Since splitCommand and remainingCommand are both array of Strings,
the coding standard specifies them to have a plural naming.

Some variable names are also made clearer, such as from ret to message.

RussellDash332 committed:
Merge branch 'branch-A-CodingStandard'

RussellDash332 committed:
[Level-9] Find

To handle this new command, a FindCommand class is added, as well as the
methods to preprocess the command.

Ui and TaskList are also modified slightly to incorporate the state of a
filtered task list.

RussellDash332 committed:
Merge branch 'branch-Level-9'

This commit merges branch-Level-9 to the master branch and at the same time
resolves the existing conflict. This is due to the updates on the coding standard
that affect some of the variable name changes.

RussellDash332 committed:
Fix bug on method rename

RussellDash332 committed:
Add Checkstyle build config

Taken from se-edu/addressbook-level3

RussellDash332 committed:
[A-CheckStyle] Use CheckStyle

Resolve all CheckStyle issues, such as incorrect import ordering,
wrong position of '+', and newline at the end of file.

RussellDash332 committed:
Change semicolon to a Java comment

darrentde commented on src/main/java/stashy/parser/Parser.java:
```diff
@@ -0,0 +1,317 @@
+                + "(?<by>(by: [^\\(\\)]*)?)"
+                + "(?<at>(at: [^\\(\\)]*)?)\\)?$");
+
+    public static final String[] ACCEPTABLE_DATETIME_FORMATS = {
```
I like how you accept multiple multiple formats for reading date time and extracted them into a constant array👍

RussellDash332 replied:
Thanks! I feel this is an ad-hoc solution to detect multiple date formats but it actually makes the parser more readable 😅

darrentde reviewed (commented):
LGTM. Your code looks really clean and I didn't find any violations of Java coding standards. Good job 🤩

RussellDash332 committed:
Modify build.gradle and gitignore

RussellDash332 committed:
Modify JavaFX config on build.gradle

RussellDash332 committed:
Resolve CheckStyle issue on JUnit classes

RussellDash332 committed:
First iteration of JavaFX integration

RussellDash332 committed:
Second and third iteration of JavaFX integration

RussellDash332 committed:
Add FXML boilerplate code

RussellDash332 committed:
Minor updates

Add image resources and fix the main window FXML

RussellDash332 committed:
Modify gitignore

RussellDash332 committed:
Change images

RussellDash332 committed:
[Level-10] GUI

- Finalize GUI integration for Stashy by modifying Launcher and Ui
- Add all String-returning version of the execute methods for each Command classes
- Slight modification to the toString() method of TaskList
- Better naming for the LocalDateTime variables in both Deadline and Event
- Customize FXML for both DialogBox and MainWindow

RussellDash332 committed:
Merge branch 'branch-Level-10'
GUI integration using JavaFX and FXML

RussellDash332 committed:
Change DateTimeFormatter as a static attribute in Task

The byDateTime and atDateTime attributes in Deadline and Event,
respectively, have a common usage DateTimeFormatter, which is
to output the LocalDateTime attribute as a formatted String.

This is an instance of code duplication.

Let's put the result of the DateTimeFormatter.ofPattern method in
Task as a static final attribute called OUTPUT_DATETIME_PATTERN.

This will not only reduce code duplication, but also allow other
subclasses of Task to handle datetime-related features easier in
the future. This includes but not limited to:
* storing the datetime the task is marked or unmarked
* storing the datetime the task is created or modified
* checking for datetime validity

RussellDash332 committed:
[A-FullCommitMessage] Change datetime pattern for output

The datetime pattern for outputting Task objects as a String can be
ambiguous at times.

The year and time of the event can have the exact same value, for
example, 20:22 at the year 2022.

By adding a colon between the hours and the minutes, there should be
a clear distinction between the year and the time of a datetime string.

Let's add a single colon on the static final attribute called
OUTPUT_DATETIME_PATTERN and wrap to a new line to avoid possible
CheckStyle errors.

RussellDash332 committed:
Modify comment on catch block

RussellDash332 committed:
[A-CodeQuality] Remove unused overloaded method

RussellDash332 committed:
Bugfix at FindCommand

RussellDash332 committed:
Make FindCommand case-insensitive

RussellDash332 committed:
[A-Assertions] Use Assertions

Assert statements can be used at some parts of the code to double
check on trivial things such as non-negative task ID, and using regex
to capture the strings on each line of an existing task data file.

Let's put some basic assert statements that might be useful when giving
out-of-the-box inputs as an user, this includes negative task ID input
and modifying the local task data file manually.

RussellDash332 committed:
[A-Streams] Add usage of Java streams

RussellDash332 committed:
Merge pull request #1 from RussellDash332/branch-A-Assertions

A-Assertions

RussellDash332 committed:
Merge pull request #2 from RussellDash332/branch-A-CodeQuality

A-CodeQuality

RussellDash332 committed:
Merge branch 'master' into branch-A-Streams

RussellDash332 committed:
Merge pull request #3 from RussellDash332/branch-A-Streams

A-Streams

RussellDash332 committed:
[A-CI] Set up CI

RussellDash332 committed:
Fix CI failure bugs

RussellDash332 committed:
Fix incorrect CLI commands for text UI testing

RussellDash332 committed:
Add Javadoc comments and modify Stashy-Ui interaction

Some of the Javadoc comments on non-void methods have no @return tag.

Ui should have a method that shows the error message as a String instead
of a void. This way, the additional text can be shown in the GUI as well
if there is any StashyException caught during the command execution.

RussellDash332 committed:
Update text UI testing

RussellDash332 committed:
Allow reproducibility on text UI testing

marcuslowhuiyu commented on src/main/java/stashy/parser/Parser.java:
```diff
@@ -0,0 +1,325 @@
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+
+import stashy.commands.AddDeadlineCommand;
```
I like how you import each class, instead of importing the whole package through *

marcuslowhuiyu commented on src/main/java/stashy/data/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+ * Represents a task.
+ */
+public abstract class Task {
+    protected static final DateTimeFormatter OUTPUT_DATETIME_PATTERN =
```
I like how you use all uppercases to name your constant

marcuslowhuiyu commented on src/main/java/stashy/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,60 @@
+ * A Command specifically designed to add an Event object.
+ */
+public class AddEventCommand extends Command {
+    public static final String KEYWORD = "event";
```
Is there any reason for this to be publicly accessible?

RussellDash332 replied:
It could have been `protected`, thanks for highlighting this!

marcuslowhuiyu commented on src/main/java/stashy/launcher/Main.java:
```diff
@@ -0,0 +1,34 @@
+            fxmlLoader.<MainWindow>getController().setStashy(stashy);
+            stage.show();
+        } catch (IOException e) {
+            e.printStackTrace();
```
I think you can considering throwing your own exception here instead of printStackTrace

RussellDash332 replied:
I agree, this is still taken from the example Main.java given in the JavaFX tutorial 😅

RussellDash332 committed:
Update text UI testing

RussellDash332 committed:
[C-Help] Give help to users

RussellDash332 committed:
Modify EXPECTED.TXT

RussellDash332 committed:
Merge pull request #4 from RussellDash332/branch-C-Help

This extension allows users to be guided to the list of available commands.

Doing the 'help <command>' command will guide the users to the guide
for that specific command.

Let's add the help command to support this extension.

RussellDash332 committed:
Change Stashy GUI version

RussellDash332 committed:
Fix minor UI issues

There were some minor issues that might cause ambiguity.
    - ExitCommand: example command should be 'bye' instead of 'exit'
    - HelpCommand: Last sentence of help page should be included upon start
    - Ui: Put the last sentence from HelpCommand on showCommandList()
    - Update text UI testing

Let's update them.

RussellDash332 committed:
[A-BetterGui] Better GUI

RussellDash332 committed:
[A-Personality] A unique personality

RussellDash332 committed:
[A-MoreErrorHandling] More error handling

RussellDash332 committed:
[A-MoreTesting] More testing

RussellDash332 committed:
Fix Javadoc comments and boolean naming

RussellDash332 committed:
[A-UserGuide] User Guide

RussellDash332 committed:
[A-Release] Release

RussellDash332 committed:
Update README.md and version

RussellDash332 committed:
Update Ui.png

RussellDash332 committed:
Update user guide

RussellDash332 committed:
Exit platform when chat ended

RussellDash332 committed:
Update user guide

RussellDash332 committed:
Add newline to user guide
